### PR TITLE
Add Support For MongoDB Client Side Field Level Encryption (CSFLE)

### PR DESCRIPTION
**Pull request by dill0wn**
This implementation was guided largely by [pymongo's examples on explicit, manual encryption](https://pymongo.readthedocs.io/en/stable/examples/encryption.html#explicit-encryption-with-automatic-decryption).

These changes add a lot:
- The `Document` class now supports automatic encryption and decryption of individual fields. It gains this ability by inheriting from `ming.encryption.EncryptedDocumentMixin` and leveraging the new quasi-field class `ming.encryption.DecryptedField`.
- Encryption configuration options are handled by `ming.encryption.EncryptionConfig`
    - `ming.configure` automatically parses flattened config values like you would see in an ini. For example:
    
    ```
    ming.main.encryption.kms_providers.local.key = ENCR_KEY
    ming.main.encryption.key_vault_namespace = encryption_test.dataKeyVault
    ming.main.encryption.provider_options.local.key_alt_names = ["datakey_test1"]
    ```
- A new formencode validator `ming.validators.EncryptionConfigValidator` has been added to validate config dict values.
- For runtime usage, the `EncryptionConfig` is added to the `DataStore` instance as a new `DataStore.encryption` instance property. This is what individual `Document`s reference when performing encryption/decryption.
- Added new package dependencies: `pymongo[decryption]` and `cachetools`
- Added tests for new encryption features
- Added new demo `docs/presentations/demo_encryption.py`  

  

See pymongo's documentation on encryption for further details on the implementation and expected configuration.
- https://pymongo.readthedocs.io/en/stable/examples/encryption.html
- https://pymongo.readthedocs.io/en/stable/api/pymongo/encryption.html

### CODE REVIEW

**Comment by dill0wn** [comment]
This work is not ready to merge yet as it only covers Documents. We still need to support MappedClasses.

**Review by brondsem — CHANGES_REQUESTED** [review]

File: ming/encryption.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    def __set__(self, instance: EncryptedDocumentMixin, value: T):
+        if not isinstance(value, self.field_type):
+            raise TypeError(f'not {self.field_type}, got {value!r}')
+        setattr(instance, self.encrypted_field, instance.encr(value))
+
+
+class EncryptedDocumentMixin:
+
+    @classmethod
+    @cached(RRCache(maxsize=99))  # needs to be per datastore, so we pass that as a param
```
[the remark this answers is not in the document]
Comment (dill0wn) [reply ming/encryption.py:72]: dropped cachetools and used functools.cache

File: ming/validators.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                        error_dict['kms_providers'] = validators.Invalid(self.message('InvalidKMSProvider', state, provider=key), config, state)
+                        break
+                    used_kms_providers.add(key)
+                
+                if error_dict:
+                    return error_dict
+                
+                if 'local' in used_kms_providers:
+                    if 'key' not in config.kms_providers['local']:
+                        error_dict['kms_providers'] = validators.Invalid(f'Local KMS Provider requires a key', config, state)
```
[the remark this answers is not in the document]
Comment (dill0wn) [reply ming/validators.py:52]: Greatly changed how this all works. It's both more clear and less clear. The goal was to try and round up as many validation errors in one go as possible. I may have overlooked something here though. Hopefully the latest changes are an improvement

File: ming/validators.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        if not config:
+            return # no encryption settings
+
+        def validate_inner():
+            error_dict = {}
+            if config.kms_providers:
+                if not config.key_vault_namespace:
+                    error_dict['key_vault_namespace'] = validators.Invalid(self.message('InvalidKMSProvider', state), config, state)
+                if not config.provider_options:
+                    error_dict['provider_options'] = validators.Invalid(self.message('InvalidKMSProvider', state), config, state)
```
[the remark this answers is not in the document]
Comment (dill0wn) [reply ming/validators.py:38]: Big overhaul here. Hopefully it's ok.

File: ming/validators.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    def _validate_python(self, config: EncryptionConfig, state):
+        if not config:
+            return # no encryption settings
+
+        def validate_inner():
+            error_dict = {}
+            if config.kms_providers:
+                if not config.key_vault_namespace:
+                    error_dict['key_vault_namespace'] = validators.Invalid(self.message('InvalidKMSProvider', state), config, state)
```
[the remark this answers is not in the document]
Comment (dill0wn) [reply ming/validators.py:36]: Big overhaul here, hopefully it's better.

File: ming/validators.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+try:
+    from formencode import schema, validators
+except ImportError:
+    raise MingConfigError("Need to install FormEncode to use ``ming.encryption`` package")
+
+
+class EncryptionConfigValidator(validators.FancyValidator):
+    """
+    Password string validation to refrain from usage of name, username or email id (full or partial) as its substrings
```
[the remark this answers is not in the document]
Comment (dill0wn) [reply ming/validators.py:13]: Whoops, good catch. updated

File: ming/encryption.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    def decrypt_some_fields(self) -> dict:
+        # useful for json, removes encrypted fields and uses decrypted forms
+        decrypted_data = dict(self)
+        for k, v in self.items():
+            if k.endswith('_encrypted'):
+                del decrypted_data[k]
+                k_decrypted = k.replace('_encrypted', '')
+                decrypted_data[k_decrypted] = getattr(self, k_decrypted)
+        return decrypted_data
```
[the remark this answers is not in the document]
Comment (dill0wn) [reply ming/encryption.py:144]: Added coverage (and updated .pyi definitions)

File: docs/encryption.rst [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        email_encrypted = Field(S.Binary, if_missing=None)
+        email = DecryptedField(str, 'email_encrypted')
+
+
+Breaking it Down
+========================
+
+This approach requires that you follow a few conventions in order to function correctly.
+
+.. 1. Fields  encrypted data must be named with the suffix `_encrypted`.
```
Comment (brondsem) [inline docs/encryption.rst:38]: this line doesn't show up when I generate the HTML

File: ming/datastore.py [hunk 8]
```diff
@@ -191,3 +203,53 @@ def db(self) -> Database:
...
+
+    @property
+    def encryption(self) -> encryption.EncryptionConfig | None:
+        return self._encryption_config
+
+    @LazyProperty
+    def encryptor(self) -> ClientEncryption:
+        """Creates and returns a :class:`pymongo.encryption.ClientEncryption` instance for the given ming datastore. It uses this to handle encryption/decryption using pymongo's native routines.
+
+        :param ming_ds: the :class:`ming.datastore.Datastore` for which this encryptor should be configured with.
```
Comment (brondsem) [inline ming/datastore.py:215]: can delete this line, since `self` is the datastore now!

File: ming/odm/property.py
Comment (brondsem) [inline ming/odm/property.py:1]: in `def include_in_repr(self):` lets add Binary types to the skip list.  I tested `repr(u)` in one of the encryption tests to see what it'd look like.  I don't think we want binary!

File: setup.py [hunk 9]
```diff
@@ -34,7 +34,7 @@
       include_package_data=True,
       zip_safe=True,
       install_requires=[
-        "pymongo",
+        "pymongo[encryption]<4.9",
```
Comment (brondsem) [inline setup.py:37]: I'm ok with this for now, to get this MR merged.  But we probably should make it work on 4.9 next, before making a release :|

**Review by brondsem — CHANGES_REQUESTED** [review]

File: setup.py (hunk 9 above)
Comment (dill0wn) [reply setup.py:37]: Agreed.

File: ming/datastore.py (hunk 8 above)
Comment (dill0wn) [reply ming/datastore.py]: Fixed

**Review by brondsem — APPROVED** [review]

**Pull request merged**